cpp/oneapi/dal/algo/covariance/backend/gpu/compute_kernel_dense_dpc.cpp

```diff
@@ -14,117 +14,136 @@
+    auto& q = ctx.get_queue();
+    const auto data = input.get_data();
 
-    auto arr_means = array<Float>::empty(queue, component_count, sycl::usm::alloc::device);
-    const auto daal_means = interop::convert_to_daal_table(queue, arr_means, 1, component_count);
+    const std::int64_t row_count = data.get_row_count();
+    const std::int64_t column_count = data.get_column_count();
+    const std::int64_t component_count = data.get_column_count();
+    dal::detail::check_mul_overflow(row_count, column_count);
+    dal::detail::check_mul_overflow(column_count, column_count);
+    dal::detail::check_mul_overflow(component_count, column_count);
 
-    auto result = compute_result<Task>{}.set_result_options(desc.get_result_options());
+    const auto data_nd = pr::table2ndarray<Float>(q, data, sycl::usm::alloc::device);
 
-    if (desc.get_result_options().test(result_options::cov_matrix)) {
-        auto arr_cov_matrix =
-            array<Float>::empty(queue, component_count * component_count, sycl::usm::alloc::device);
-        const auto daal_cov_matrix =
-            interop::convert_to_daal_table(queue, arr_cov_matrix, component_count, component_count);
-
-        interop::status_to_exception(
-            daal_covariance_kernel_t<Float>().compute(daal_data.get(),
-                                                      daal_cov_matrix.get(),
-                                                      daal_means.get(),
-                                                      &daal_parameter));
-        is_mean_computed = true;
+    auto [sums, sums_event] = compute_sums(q, data_nd);
 
+    if (desc.get_result_options().test(result_options::cov_matrix)) {
+        auto [cov, means, vars, cov_event] = compute_covariance(q, data_nd, sums, { sums_event });
+        //is_mean_computed = true;
+        //cov_event.wait_and_throw();
         result.set_cov_matrix(
-            homogen_table::wrap(arr_cov_matrix, component_count, component_count));
+            (homogen_table::wrap(cov.flatten(q, { cov_event }), column_count, column_count)));
     }
     if (desc.get_result_options().test(result_options::cor_matrix)) {
-        auto arr_cor_matrix =
-            array<Float>::empty(queue, component_count * component_count, sycl::usm::alloc::device);
-        const auto daal_cor_matrix =
-            interop::convert_to_daal_table(queue, arr_cor_matrix, component_count, component_count);
-        daal_parameter.outputMatrixType = daal_covariance::correlationMatrix;
-
-        interop::status_to_exception(
-            daal_covariance_kernel_t<Float>().compute(daal_data.get(),
-                                                      daal_cor_matrix.get(),
-                                                      daal_means.get(),
-                                                      &daal_parameter));
-        is_mean_computed = true;
+        auto [corr, means, vars, corr_event] =
+            compute_correlation(q, data_nd, sums, { sums_event });
```

Comment:
We can utilize results of the covariance computation if it have already be done. Please optimize it via reusing